### PR TITLE
[VERSIONING.md] Reorganize and clarify compatibility guarantees

### LINKED ISSUE #8002: Document Go API stability guarantees
To start marking more modules as stable, we want to discuss the following open questions about configuration structs:

1. Does adding new validation rules mean a breaking change?
2. Should we offer a "NewDefault...." for each config and say that the behavior of the config is stable only if initialized with NewDefault?
3. Is adding new fields means breaking change? Let's assume someone "squash" the message into another message, then adding a field with same mapstruct value will be a breaking change, what do we do with that? What are the rules we follow.

We generally also need to think of any special implications for configs that are not defined by the golang conventions of breaking compatibility.

_Reworded from post by @bogdandrutu in https://github.com/open-telemetry/opentelemetry-collector/issues/7954#issuecomment-1601902350_

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** Reorganize `VERSIONING.md`, specify compatibility guarantees that have already been discussed and answer questions on #8002.

Three new guarantees in the document had already been discussed elsewhere:

- String representation: https://github.com/open-telemetry/opentelemetry-collector/issues/7625#issuecomment-1674789807
- Go version compatibility: Mentioned [on README.md](https://github.com/open-telemetry/opentelemetry-collector?tab=readme-ov-file#compatibility)
- Dependency updates: Discussed in private, most recently in relation to whether #7095 should block `pdata`'s 1.0 (consensus seems to be that it should not).

Regarding the questions mentioned on #8002:

> Does adding new validation rules mean a breaking change?

Generally, yes except when the configuration was already invalid (i.e. the Collector did not work properly with it).

> Should we offer a "NewDefault...." for each config and say that the behavior of the config is stable only if initialized with NewDefault?

I did not add anything for this one. I think we should discuss it, but:
- I don't think there is a sensible output for `NewDefault...` for all configurations (e.g. for `confignet.TCPAddr`, what would the default be?)
- It seems to me that we should handle configuration validity through `Validate`, and not through `NewDefault...`. `NewDefault` may help but ultimately we need to verify through `Validate`.

> Is adding new fields means breaking change? Let's assume someone "squash" the message into another message, then adding a field with same mapstruct value will be a breaking change, what do we do with that? What are the rules we follow.

This was already in this document, where we had decided that adding new fields was okay. I think it would be too stringent to bar us from adding new fields.

**Link to tracking Issue:** Fixes #8002

### CODE REVIEW

File: VERSIONING.md [hunk 1]
```diff
@@ -72,14 +107,8 @@ is designed so that the following goal can be achieved:
           whenever you are using the module name).
         * If a module is version `v0` or `v1`, do not include the major version
           in either the module path or the import path.
-    * Configuration structures should be considered part of the public API and backward
-      compatibility maintained through any changes made to configuration structures.
-        * Because configuration structures are typically instantiated through unmarshalling
-          a serialized representation of the structure, and not through structure literals,
-          additive changes to the set of exported fields in a configuration structure are
-          not considered to break backward compatibility.
     * Modules will be used to encapsulate receivers, processor, exporters,
```
Comment (songy23) [inline VERSIONING.md:110]:
```suggestion
    * Modules will be used to encapsulate receivers, processors, exporters,
```

File: VERSIONING.md [hunk 2]
```diff
@@ -5,7 +5,52 @@ is designed so that the following goal can be achieved:
...
+  break the struct by creating a conflict with an existing method of the other embedded type. We cannot protect 
+  against this rare case and do not guarantee compatibility in such scenarios.
+* **Dot imports.** If a program imports a package using `import .`, additional names defined in the imported package
+  in future releases may conflict with other names defined in the program. We do not recommend the use of
+  `import .` with OpenTelemetry Collector modules.
+
+Unless otherwise specified in the documentation, the following may change in any way between minor versions:
+* **String representation**. The `String` method of any struct is intended to be human-readable and may change its output
+  in any way.
+* **Go version compatibility**. Removing support for an unsupported Go version is not considered a breaking change.
```
Comment (songy23) [inline VERSIONING.md:31]: How about OS compatibility?

File: VERSIONING.md [hunk 3]
```diff
@@ -5,7 +5,52 @@ is designed so that the following goal can be achieved:
...
+Unless otherwise specified in the documentation, the following may change in any way between minor versions:
+* **Adding new fields to configuration structures**. Because configuration structures are typically instantiated through 
+unmarshalling a serialized representation of the structure, and not through structure literals, additive changes to 
+the set of exported fields in a configuration structure are not considered to break backward compatibility.
+
+The following are explicitly considered to be breaking changes:
+* **Modifying struct tags related to serialization**. Struct tags used to configure serialization mechanisms (`yaml:`, 
+`mapstructure:`, etc) are part of the structure definition and must maintain compatibility to the same extent as the 
+structure.
+* **Making validation rules more strict**. A valid configuration struct as defined by its `Validate` method return value
```
Comment (songy23) [inline VERSIONING.md:49]: Do you want to mention the opposite i.e. relaxing validation rules is considered backwards-compatible?

File: VERSIONING.md (hunk 2 above)
Comment (mx-psi) [reply VERSIONING.md:31]: Added a bullet point for this

File: VERSIONING.md (hunk 3 above)
Comment (mx-psi) [reply VERSIONING.md:52]: Added it, this one seemed a bit more obvious to me, but I guess it doesn't hurt to add it

**Review by songy23 — APPROVED** [review]
LGTM

**Review by atoulme — APPROVED** [review]

File: VERSIONING.md [hunk 4]
```diff
@@ -5,7 +5,55 @@ is designed so that the following goal can be achieved:
...
+  break the struct by creating a conflict with an existing method of the other embedded type. We cannot protect 
+  against this rare case and do not guarantee compatibility in such scenarios.
+* **Dot imports.** If a program imports a package using `import .`, additional names defined in the imported package
+  in future releases may conflict with other names defined in the program. We do not recommend the use of
+  `import .` with OpenTelemetry Collector modules.
+
+Unless otherwise specified in the documentation, the following may change in any way between minor versions:
+* **String representation**. The `String` method of any struct is intended to be human-readable and may change its output
+  in any way.
+* **Go version compatibility**. Removing support for an unsupported Go version is not considered a breaking change.
```
[the remark this answers is not in the document]
Comment (mx-psi) [reply VERSIONING.md:31]: Supported by the Go team. This is mostly a rephrasing of what is already on the README https://github.com/open-telemetry/opentelemetry-collector?tab=readme-ov-file#compatibility

**Review by TylerHelmuth — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]
LGTM. Thanks for putting this together.

**Pull request merged**